### PR TITLE
Meeting scheduling functionality

### PR DESCRIPTION
**Pull request by jaimehisao**
There is a lot of debug output. This will be removed in the future.

### CODE REVIEW

**Comment by jaimehisao** [comment]
Initial draft of the PR. Comments are a work in progress and also the algorithm isn't optimized yet. Focusing on having a working solution and then optimize the code.

**Review by ssmall — COMMENTED** [review]
Just leaving a comment to set review state -- re-request review when it's ready with the latest changes

**Comment by jaimehisao** [comment]
Changed from having a "split" function to having an array with every minute of the day.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:17]: [nit] duplicate import

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: Nice use of Java doc features! Could you add the big-O runtime of your algorithm too?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: [nit] here and in the rest of this file, leave a space between comments "//" and start of comment: https://google.github.io/styleguide/javaguide.html#s4.8.6-comments

**Review by ssmall — CHANGES_REQUESTED** [review]
This PR doesn't appear to include any changes to unit tests -- please include the additional unit tests described in https://github.com/jaimehisao/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md#writing-more-tests

**Comment by jaimehisao** [comment]
Yes, I'm working on those right now.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: [nit] *attendees, and on L59.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: A couple of performance notes here, a call to remove an object from an ArrayList is going to cause an `O(n)` search for the object and an `O(n)` removal for random deletion from an array.

For best performance on random deletion, you probably want a LinkedList which gives `O(1)` removal.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I'm not sure I understand the goal here.

Because of the `!availableTimes[i]` check, it seems like `intervalDuration` is supposed to represent a period of **unavailable** times, but then these are added to the `possibleTimes` array?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: How do TimeRanges where there are no `events` at all get marked as `true`? It seems like if there's no `event` during a TimeRange, it should be marked as available by default?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think you can actually omit this check and removing from localEvents, since you can actually detect this scenario while iterating over `availableTimes`. Ideally, in the case where there are no Events where eventAttendees and mandatoryAttendees overlap, the result of `possibleTimes` after iterating over `availableTimes` is one TimeRange equivalent to WHOLE_DAY.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Does `availableTimes` actually end up being an array where all available times are marked with `true`?

I'm not certain that's true and I think it's a core piece of understanding what the algorithm is doing.

**Review by anniefu — COMMENTED** [review]
Good use of comments!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (jaimehisao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I was actually marking the busy times. It was available times at first but then I marked the times with meetings.
Will rename accordingly.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (jaimehisao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:18]: actually, when I remove either of those I get a compilation error...

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (jaimehisao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: availableTimes holds true where there's a meeting. Wrong choice of name, which has been fixed.

**Comment by jaimehisao** [comment]
Changed to now take into consideration optional attendees and added the first 3 of the required extra tests.
Also, per @anniefu 's comments, renamed the availableTimes array because it represents the times that are taken up by meetings, instead of the times available, my bad.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (jaimehisao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: removed!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (jaimehisao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes! Removing from localEvents was a temporary fix, especially for the performance issues. Neither the check, nor the removal of items exist in the newest version.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (ssmall) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:17]: one is `Collection` and the other is `Collections` ... not a duplicate, I think

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java
Comment (ssmall) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Nit: I believe this should be `B`, not `C`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (ssmall) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: instead of
```java
minuteOfTheDay < MINUTES_IN_A_DAY+1
```
you could just write 
```java
minuteOfTheDay <= MINUTES_IN_A_DAY
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (ssmall) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: Great analysis!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (ssmall) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Good use of collection utility functions!

**Review by ssmall — APPROVED** [review]
Looks great to me! Would be good to get @anniefu's approval as well before merging.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:17]: Oh, good catch. Weird that Java has both haha.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: [nit] leave spaces on around "else if": https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Since you're actually iterating over busyTimes, maybe just `minuteOfTheDay < len(busyTimes)` (or whatever the length function is in Java).

It's possible that someone in the future changes busyTimes to not be MINUTES_IN_A_DAY long.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: [nit] here and elsewhere, leave a space between "if" and open parenthesis.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: [nit] also spaces between "for" and start parenthesis.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (anniefu) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Wouldn't availabilityInterval == 1 if the last minute of the day is available? Shouldn't that be a possibleTime if my requested meeting duration is only 1 minute?

**Review by anniefu — CHANGES_REQUESTED** [review]
Nice! Makes a lot more sense now that availableTimes is busyTimes :) 

My approval is just waiting on the one question about when `availabilityInterval == 1`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (jaimehisao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yeah but the array is created with an extra "slot" so technically that 1 is accounting for that. When that `availabilityInterval == 1` isn't there a TimeRange of [1440,1440) is created. So it's a meeting that has a duration of 0

Comment (jaimehisao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I went ahead and added a test for this case. It contains a meeting that lasts from START_OF_DAY all the way to END_OF_DAY (non inclusive of 11:59PM), and the test checks if query returns a single meeting from 1439 to 1440, which is the EoD

Comment (anniefu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Cool, sounds good. Thanks for explaining and adding the test case!

**Review by anniefu — APPROVED** [review]

**Pull request merged**
